py4DSTEM/process/diffraction/crystal.py

```diff
@@ -621,7 +622,112 @@ def generate_diffraction_pattern(
+            plot_rings(bool):               if true, plot diffraction rings with plot_ring_pattern
+            return_calc (bool):             return radii and intensities 
+
+        Returns: 
+            radii_unique (np array):        radii of ring pattern in units of scattering vector k
+            intensity_unique (np array):    intensity of rings weighted by frequency of diffraciton spots
+        """ 
+       
+        if use_bloch: 
+            assert (thickness is not None), "provide thickness for dynamical diffraction calculation"
+            assert hasattr(self, "Ug_dict"), "run calculate_dynamical_structure_factors first"
+            
+        if not hasattr(self, "struct_factors"):
+            self.calculate_structure_factors(
+                k_max = k_max, 
+            )
+    
+        #check accelerating voltage 
+        if hasattr(self, "accel_voltage"): 
+            accelerating_voltage = self.accel_voltage
+        else: 
+            self.accel_voltage = 300e3
+            print("Accelerating voltage not set. Assuming 300 keV!")
+        
+        #check orientation plan
+        if not hasattr(self, "orientation_vecs"):
+            if orientation_plan_params is None: 
+                orientation_plan_params = {
+                    'zone_axis_range': 'auto',
+                    'angle_step_zone_axis': 4, 
+                    'angle_step_in_plane': 4,
+                }    
+            self.orientation_plan(
+                **orientation_plan_params,
+            )
+
+        #calculate intensity and radius for rings 
+        radii = []
+        intensity = []
+        for a0 in range(self.orientation_vecs.shape[0]):
+            if use_bloch:
+                beams = self.generate_diffraction_pattern(
+                    zone_axis_lattice = self.orientation_vecs[a0],
+                    sigma_excitation_error = sigma_excitation_error, 
+                    tol_intensity = tol_intensity, 
+                    k_max = k_max
+                )
+                pattern = self.generate_dynamical_diffraction_pattern(
+                    beams = beams,
+                    zone_axis_lattice = self.orientation_vecs[a0],
+                    thickness = thickness,
+                )
+            else:  
+                pattern = self.generate_diffraction_pattern(
+                    zone_axis_lattice = self.orientation_vecs[a0],
+                    sigma_excitation_error = sigma_excitation_error, 
+                    tol_intensity = tol_intensity, 
+                    k_max = k_max
+                )
+
+            intensity.append(pattern['intensity'])
+            radii.append((pattern['qx']**2 +  pattern['qy']**2)**0.5)
+        
+        intensity = np.concatenate(intensity)
+        radii = np.concatenate(radii)
 
+        radii_unique, index, counts, = np.unique(radii, return_counts = True, return_index = True)
+        intensity_unique = intensity[index] * counts
```

Comment:
I think that this is not summing the intensities in the right way. The `index` returned by `unique` only gives the _first_ occurrence of the value, not all of them. I think you need to use the `return_inverse` option to `unique` and then use `bincount` for the accumulation. The code below should show the problem, but I think that to make it work correctly in our case we will need to specify the `minlength` option for `bincount`...

```
In [1]: import numpy as np

In [2]: r = np.array([1,2,2,3,4,4,5])

In [3]: i = np.array([1,1,1,2,2,3,3])

In [4]: r
Out[4]: array([1, 2, 2, 3, 4, 4, 5])

In [5]: i
Out[5]: array([1, 1, 1, 2, 2, 3, 3])

In [6]: rad,idx,cts = np.unique(r, return_counts=True, return_index=True)

In [7]: i[idx]*cts
Out[7]: array([1, 2, 2, 4, 3])

In [8]: rad
Out[8]: array([1, 2, 3, 4, 5])

In [9]: idx
Out[9]: array([0, 1, 3, 4, 6])

In [10]: rad,idx,inv,cts = np.unique(r, return_counts=True, return_index=True,return_inverse=True)

In [11]: inv
Out[11]: array([0, 1, 1, 2, 3, 3, 4])

In [12]: i[inv]
Out[12]: array([1, 1, 1, 1, 2, 2, 2])

In [13]: np.bincount(inv,weights=i)
Out[13]: array([1., 2., 2., 5., 3.])
```

Reply:
Thanks for finding this error @sezelt!